File: experiments/md-embedding-server/mcp/src/envelope.js
// Response envelope for md-mcp tools.
//
// Wraps every tool reply with a uniform metadata layer so consuming skills
// always see corpus_state, lock, cost and structured next_step alongside the
// tool-specific result. Backward-compatible: the original result keys remain
// at the top level — `_envelope` is additive.
//
// Phase 1 scope (current): next_step derivation from error patterns, cost
// ledger placeholder, corpus_root resolution from args. Phase 1.2 will fill
// corpus_state via fast SQLite meta read. Phase 4 will fill lock via the
// holder-identity contract.

import { resolve } from "node:path";

const ENVELOPE_VERSION = 1;
const CORPUS_STATE_TTL_MS = 30_000;
const CORPUS_STATE_TIMEOUT_MS = 10_000;

const _ledger = {
  turn_usd: 0,
  session_usd: 0,
  turn_started_at: Date.now()
};

export function recordCost(usd) {
  if (typeof usd !== "number" || !Number.isFinite(usd) || usd <= 0) return;
  _ledger.turn_usd += usd;
  _ledger.session_usd += usd;
}

export function resetTurn() {
  _ledger.turn_usd = 0;
  _ledger.turn_started_at = Date.now();
}

export function getCostSnapshot() {
  return {
    turn_usd: round4(_ledger.turn_usd),
    session_usd: round4(_ledger.session_usd)
  };
}

function round4(n) {
  return Math.round(n * 10000) / 10000;
}

function resolveCorpusRoot(args) {
  // Only resolve from explicit corpus-shaped args. `args.path` is usually a
  // single file (md_preflight, md_impact, md_read_related), not a corpus —
  // running status on it would mis-classify state. `args.scan` is a corpus
  // scope for graph tools; `args.corpus` is the explicit navigator arg.
  if (!args || typeof args !== "object") return null;
  const candidate = args.corpus || args.scan;
  if (!candidate || typeof candidate !== "string") return null;
  try {
    return resolve(candidate);
  } catch {
    return null;
  }
}

const _corpusStateCache = new Map();

export async function quickCorpusState(corpusRoot) {
  if (!corpusRoot) return null;

  const cached = _corpusStateCache.get(corpusRoot);
  if (cached && Date.now() - cached.fetched_at < CORPUS_STATE_TTL_MS) {
    return cached.state;
  }

  try {
    const { resolveNavigatorScript } = await import("./paths.js");
    const { spawnPython, tryParseJson } = await import("./subprocess.js");
    const script = resolveNavigatorScript();
    const { code, stdout } = await spawnPython(
      script,
      ["status", corpusRoot, "--json"],
      { timeoutMs: CORPUS_STATE_TIMEOUT_MS }
    );
    // Exit 0 = success; exit 1 = "no markdown files" — both produce valid
    // state JSON. Exit 2/3 (path/dependency errors) we treat as null.
    if (code === 0 || code === 1) {
      const parsed = tryParseJson(stdout);
      if (parsed && typeof parsed === "object") {
        const slim = {
          state: parsed.state ?? null,
          model: parsed.model ?? null,
          index_exists: parsed.index_exists ?? false,
          last_touched: parsed.last_touched ?? null,
          added_sections: parsed.added_sections ?? 0,
          removed_sections: parsed.removed_sections ?? 0,
          pending_chunks: parsed.pending_chunks ?? 0,
          drift_count: parsed.drift_count ?? 0,
          metadata_mismatch: parsed.metadata_mismatch ?? false,
          delta_too_large: parsed.delta_too_large ?? false,
          recommended_action: parsed.recommended_action ?? null
        };
        _corpusStateCache.set(corpusRoot, { state: slim, fetched_at: Date.now() });
        return slim;
      }
    }
  } catch {
    // Silent fail — corpus_state stays null. We do not want a broken status
    // probe to break the underlying tool reply that the agent actually asked
    // for. Diagnose via md_ping or direct CLI.
  }
  return null;
}

export function invalidateCorpusStateCache(corpusRoot) {
  // Phase 2 mutating tools (md_index, md_init, md_strip confirm:true) should
  // call this after a successful run so the next reply sees fresh state.
  if (corpusRoot) {
    _corpusStateCache.delete(corpusRoot);
  } else {
    _corpusStateCache.clear();
  }
}

function deriveNextStep(result, { toolName, args, corpusRoot }) {
  if (!result || typeof result !== "object") return [];

  // index_warmup: tool refused to embed too many new chunks.
  // Replaces the prose self_repair string with two structured directives.
  if (result.error === "index_warmup_required") {
    const out = [];
    if (corpusRoot) {
      out.push({
        tool: "md_index",
        args: { corpus: corpusRoot, dry_run: true },
        reason: "Preview embedding cost before warming the index."
      });
      out.push({
        tool: "md_index",
        args: { corpus: corpusRoot, confirm: true },
        reason: "After reviewing the dry-run, embed pending chunks."
      });
    }
    if (toolName && args) {
      out.push({
        tool: toolName,
        args,
        reason: "Retry the original call once the index is warm."
      });
    }
    return out;
  }

  // confirm_required: mutating tool was called without confirm:true.
  // Surface the dry-run preview as the first directive.
  if (result.error === "confirm_required") {
    if (!toolName || !args) return [];
    const dryArgs = { ...args };
    delete dryArgs.confirm;
    dryArgs.dry_run = true;
    const confirmArgs = { ...args };
    delete confirmArgs.dry_run;
    confirmArgs.confirm = true;
    return [
      {
        tool: toolName,
        args: dryArgs,
        reason: "Preview affected files before mutation."
      },
      {
        tool: toolName,
        args: confirmArgs,
        reason: "After reviewing the dry-run, apply with confirm:true."
      }
    ];
  }

  // empty: tool ran but returned nothing useful. Hint at scope widening.
  if (result.empty === true) {
    if (toolName !== "md_search" || !args) return [];
    const broaderArgs = { ...args, scope: "descriptions" };
    return [
      {
        tool: "md_search",
        args: broaderArgs,
        reason: "Retry with scope='descriptions' for higher-level matching."
      }
    ];
  }

  return [];
}

// Fields the agent typically iterates. Counting them across known names lets
// us report items_returned without per-tool knowledge.
const COUNTABLE_LIST_FIELDS = [
  "results", "files", "items", "pairs", "proposals", "sections",
  "concepts", "headings", "cycles", "orphans", "hubs", "anchors",
  "issues", "corpora", "unindexed_with_md", "must_read", "must_update",
  "cascade_breaks", "reference_breaks", "body_wikilink_refs",
  "body_markdown_refs", "modified", "changes", "topics", "candidates",
  "folder_breakdown", "scopes"
];

// Soft threshold for "this reply is heavy on attention budget". Surfaces a
// warning so the agent can switch to a narrower path (`top: 5`, scope,
// path_include, or future verbose:false default).
const LARGE_REPLY_BYTES = 10_000;

function computeSizeEstimate(result) {
  if (!result || typeof result !== "object") return null;

  let bytes = null;
  try {
    bytes = JSON.stringify(result).length;
  } catch {}

  let items_returned = 0;
  let counted_fields = [];
  for (const field of COUNTABLE_LIST_FIELDS) {
    const v = result[field];
    if (Array.isArray(v) && v.length > 0) {
      items_returned += v.length;
      counted_fields.push({ field, count: v.length });
    }
  }

  const out = { bytes, items_returned };
  if (counted_fields.length > 0) out.counted_fields = counted_fields;
  if (bytes !== null && bytes > LARGE_REPLY_BYTES) {
    out.large_reply = true;
  }
  return out;
}

/**
 * Wrap a tool result in the standard md-mcp envelope.
 *
 * The original result keys are preserved at the top level so existing callers
 * see no change. The `_envelope` key is additive metadata for newer consumers.
 *
 * @param {*} result - The tool's raw return value (object, string or null).
 * @param {object} options
 * @param {string} options.toolName - The md_* tool name (for next_step).
 * @param {object} [options.args] - Original tool args (for next_step retries).
 * @param {object} [options.lock] - Lock holder info (Phase 4).
 * @param {object} [options.corpusState] - Index/model state (Phase 1.2).
 * @returns {object} Wrapped result with `_envelope` field.
 */
export function wrap(result, { toolName, args, lock = null, corpusState = null } = {}) {
  const corpusRoot = resolveCorpusRoot(args);
  const nextStep = deriveNextStep(result, { toolName, args, corpusRoot });
  const size_estimate = computeSizeEstimate(result);

  const envelope = {
    version: ENVELOPE_VERSION,
    tool: toolName || null,
    corpus_root: corpusRoot,
    corpus_state: corpusState,
    lock,
    cost: getCostSnapshot(),
    size_estimate,
    next_step: nextStep
  };

  if (result === null || result === undefined) {
    return { _envelope: envelope };
  }
  if (typeof result !== "object" || Array.isArray(result)) {
    return { value: result, _envelope: envelope };
  }
  if ("_envelope" in result) {
    // Already wrapped (composite case); merge instead of double-wrapping.
    return result;
  }
  return { ...result, _envelope: envelope };
}

File: experiments/md-embedding-server/src/md_cli/envelope.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from .cost_ledger import get_cost_snapshot
from .next_steps import LARGE_REPLY_BYTES, derive_next_step


ENVELOPE_VERSION = 1
COUNTABLE_LIST_FIELDS = (
    "results",
    "files",
    "items",
    "blocks",
    "pairs",
    "proposals",
    "sections",
    "concepts",
    "headings",
    "cycles",
    "orphans",
    "hubs",
    "anchors",
    "issues",
    "corpora",
    "unindexed_with_md",
    "must_read",
    "must_update",
    "cascade_breaks",
    "reference_breaks",
    "body_wikilink_refs",
    "body_markdown_refs",
    "modified",
    "changes",
    "topics",
    "candidates",
    "folder_breakdown",
    "scopes",
    "folders",
    "start_here",
)

# --- Agent-view projection -------------------------------------------------
# Internal/engine fields no agent consumes; dropped from every payload before
# it reaches the wire. NOT applied to `_envelope` (it is attached after wrap).
INTERNAL_FIELDS = frozenset({"rowid", "content_hash"})
# Path-valued keys to relativize against corpus_root for compactness.
PATH_STRING_FIELDS = frozenset({"path", "centroid_path"})
PATH_LIST_FIELDS = frozenset({"affected_files", "files_to_modify"})
# Anchor paths stay absolute (skills resolve other paths against them).
ANCHOR_PATH_FIELDS = frozenset({"root", "corpus_root", "corpus", "index_path", "scan", "repo_root"})


def _relativize(value: str, corpus_root: str) -> str:
    try:
        p = Path(value)
        if p.is_absolute():
            return str(p.relative_to(corpus_root))
    except (ValueError, TypeError, OSError):
        pass
    return value


def _project_node(node: Any, corpus_root: str | None) -> Any:
    if isinstance(node, dict):
        out: dict[str, Any] = {}
        for key, val in node.items():
            if key in INTERNAL_FIELDS:
                continue
            if corpus_root and key in PATH_STRING_FIELDS and isinstance(val, str) and val:
                out[key] = _relativize(val, corpus_root)
            elif corpus_root and key in PATH_LIST_FIELDS and isinstance(val, list):
                out[key] = [
                    _relativize(v, corpus_root) if isinstance(v, str) else _project_node(v, corpus_root)
                    for v in val
                ]
            else:
                out[key] = _project_node(val, corpus_root)
        return out
    if isinstance(node, list):
        return [_project_node(item, corpus_root) for item in node]
    return node


def project_payload(result: Any, *, tool_name: str | None = None, corpus_root: str | None = None) -> Any:
    """Agent-view projection: drop internal fields, relativize non-anchor paths,
    collapse the legacy view-flag trio (map_only/content_included/expanded) into a
    single `expanded` bool + derived `view`. Pure dict->dict; runs inside wrap()
    before the size estimate, so size reflects the projected payload. Never sees
    `_envelope` (attached after wrap), so `recommended_action` is untouched."""
    if not isinstance(result, dict):
        return result
    projected = _project_node(result, corpus_root)
    if "map_only" in projected or "content_included" in projected or "expanded" in projected:
        expanded = bool(projected.get("expanded"))
        projected.pop("map_only", None)
        projected.pop("content_included", None)
        projected["expanded"] = expanded
        projected.setdefault("view", "expanded" if expanded else "map")
    return projected


def _as_args_dict(args: dict[str, Any] | Any | None) -> dict[str, Any]:
    if args is None:
        return {}
    if isinstance(args, dict):
        raw = dict(args)
    else:
        raw = vars(args)
    skip = {"subcommand", "json"}
    return {
        key: _jsonable(value)
        for key, value in raw.items()
        if not key.startswith("_") and key not in skip and value not in (None, False)
    }


def _jsonable(value: Any) -> Any:
    if isinstance(value, Path):
        return str(value)
    if isinstance(value, (str, int, float, bool)) or value is None:
        return value
    if isinstance(value, (list, tuple)):
        return [_jsonable(item) for item in value]
    if isinstance(value, dict):
        return {str(key): _jsonable(item) for key, item in value.items()}
    return str(value)


def resolve_corpus_root(args: dict[str, Any] | Any | None) -> str | None:
    args_dict = _as_args_dict(args)
    candidate = args_dict.get("corpus") or args_dict.get("scan")
    if not isinstance(candidate, str) or not candidate:
        return None
    return str(Path(candidate).expanduser().resolve())


def compute_size_estimate(result: Any) -> dict[str, Any] | None:
    if not isinstance(result, dict):
        return None
    try:
        byte_count = len(json.dumps(result, ensure_ascii=False))
    except (TypeError, ValueError):
        byte_count = None
    items_returned = 0
    counted_fields = []
    for field in COUNTABLE_LIST_FIELDS:
        value = result.get(field)
        if isinstance(value, list) and value:
            items_returned += len(value)
            counted_fields.append({"field": field, "count": len(value)})
    estimate: dict[str, Any] = {
        "bytes": byte_count,
        "items_returned": items_returned,
    }
    if counted_fields:
        estimate["counted_fields"] = counted_fields
    if byte_count is not None and byte_count > LARGE_REPLY_BYTES:
        estimate["large_reply"] = True
    return estimate


def wrap(
    result: Any,
    *,
    tool_name: str | None,
    args: dict[str, Any] | Any | None,
    corpus_state: dict[str, Any] | None = None,
    lock: dict[str, Any] | None = None,
) -> dict[str, Any]:
    if isinstance(result, dict) and "_envelope" in result:
        return result
    args_dict = _as_args_dict(args)
    corpus_root = resolve_corpus_root(args_dict)
    result = project_payload(result, tool_name=tool_name, corpus_root=corpus_root)
    size_estimate = compute_size_estimate(result)
    envelope = {
        "version": ENVELOPE_VERSION,
        "tool": tool_name,
        "corpus_root": corpus_root,
        "corpus_state": corpus_state,
        "lock": lock,
        "cost": get_cost_snapshot(),
        "size_estimate": size_estimate,
        "next_step": derive_next_step(
            result,
            tool_name=tool_name,
            args_dict=args_dict,
            corpus_root=corpus_root,
            corpus_state=corpus_state,
            lock=lock,
            size_estimate=size_estimate,
        ),
    }
    if result is None:
        return {"_envelope": envelope}
    if not isinstance(result, dict):
        return {"value": result, "_envelope": envelope}
    return {**result, "_envelope": envelope}
